11-Backbonejs/06-Adding-Custom-Events.md

```diff
@@ -0,0 +1,255 @@
+
+Looking at the TaskViews in the Codepen, we can see when a task is clicked on using a DOM event.  However it would be REALLY useful if we could update another view when a Todo item is clicked on, for example to show more details on the selected Todo item.  Thus we will return to the Backbone Events API and the idea of an Event Bus later.
+
+### listenTo 
+
+Similar to the `.on` function, the `.listenTo()` method lets a Backbone object listen to an event that occurs on another object.  Look below and see how `listenTo()` differs from the `on()` function.  The main difference between the two is the context in which they add callbacks.  `on()` lets you directly add a callback function to another object.  `listenTo()` tells an object to listen for events on another object.  This allows the 'listener' to keep track of the objects for which it is listening.  
+
+Here's another example:
+
+```javascript
+// A view listening to a Task Model's event.
+this.listenTo(this.task, 'change:completed', this.filterOne);
+this.task.on('change:completed', this.filterOne);
+```
+
+Notice that the `listenTo()` method has the listening object subscribe to another object's event.  The on method has the model, which generates the event remember to notify the callback function.  There's another difference between `listenTo()` and `on()`.  If you call another function `stopListening` any listeners added by `listenTo()` will be removed.  Similarly if a view is removed with the `remove` method all listeners using `listenTo()` will be removed.  Models behave the same way, models deleted with the `destroy` method will call `stopListening` on any subscribers using `listenTo()`  In the **old** days, the only way to subscribe to an event was with the `on()` function.  That caused problems because when a view or model was removed the callback functions were still listening for events, which wasted memory and lead to unexpected behaviors.    
+
+The only advantage `on()` has is that with on, you can pass the context of the event.  In other words you can decide what `this` will be in the callback function.  By default, `this` in the callback function is the object that generated the event, the object who's on method you called.  However you can specify an alternative.  
+
+You can do something like this where the last argument is the context of the event handler (what `this` will refer to in the event handler).  
+
+```javascript
+// in app/views/task_list_view
+
+// prints the view to the console
+this.model.on("update", function() {
+  console.log("this is:  ", this);
+}, this);  // <--- Note the "this" 3rd parameter.
+
+// prints the view to the console
+this.listenTo(this.model, "update", function() {
+  console.log("listenTo this = ", this);
+});
+
+// prints the model to the console. 
+this.model.on("update", function() {
+  console.log("this is:  ", this);
+});
+```
+
+Honestly it's something that will *almost* never be used.
+
+So... when in doubt use `listenTo()`.  
```

Comment:
True, but since I was originally linking this to jQuery events, on made a good parallel.  I'll bring `listenTo()` up.